sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/ExpressionCodegen.scala

```diff
@@ -0,0 +1,259 @@
+          exprCodes = others
+          if (first.inputRow != null) {
+            inputRows += first.inputRow
+          }
+          first.inputVars.foreach { inputVar =>
+            if (inputVar.exprCode.code != "") {
+              exprCodes = inputVar.exprCode :: exprCodes
+            }
+          }
+        case _ =>
+      }
+    }
+    inputRows.toSeq
+  }
+
+  /**
+   * Retrieves previously evaluated columns referred by children and deferred expressions.
+   * Returned tuple contains the list of expressions and the list of generated codes.
+   */
+  def getInputVarsForChildren(
+      ctx: CodegenContext,
+      expr: Expression): (Seq[Expression], Seq[ExprCode]) = {
+    expr.children.flatMap(getInputVars(ctx, _)).distinct.unzip
+  }
+
+  /**
+   * Given a child expression, retrieves previously evaluated columns referred by it or
+   * deferred expressions which are needed to evaluate it.
+   */
+  def getInputVars(ctx: CodegenContext, child: Expression): Seq[(Expression, ExprCode)] = {
+    if (ctx.currentVars == null) {
+      return Seq.empty
+    }
+
+    child.flatMap {
+      // An evaluated variable.
+      case b @ BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null &&
+          ctx.currentVars(ordinal).code == "" =>
+        Seq((b, ctx.currentVars(ordinal)))
+
+      // An input variable which is not evaluated yet. Tracks down to find any evaluated variables
+      // in the expression path.
+      // E.g., if this expression is "d = c + 1" and "c" is not evaluated. We need to track to
+      // "c = a + b" and see if "a" and "b" are evaluated. If they are, we need to return them so
+      // to include them into parameters, if not, we track down further.
+      case b @ BoundReference(ordinal, _, _) if ctx.currentVars(ordinal) != null =>
+        trackDownVar(ctx, ctx.currentVars(ordinal))
+
+      case _ => Seq.empty
+    }.distinct
+  }
+
+  /**
+   * Tracks down previously evaluated columns referred by the generated code snippet.
+   */
+  def trackDownVar(ctx: CodegenContext, exprCode: ExprCode): Seq[(Expression, ExprCode)] = {
+    var exprCodes: List[ExprCode] = List(exprCode)
+    val inputVars = mutable.ArrayBuffer.empty[(Expression, ExprCode)]
+
+    while (exprCodes.nonEmpty) {
+      exprCodes match {
+        case first :: others =>
+          exprCodes = others
+          first.inputVars.foreach { inputVar =>
+            if (inputVar.exprCode.code == "") {
+              inputVars += ((inputVar.expr, inputVar.exprCode))
+            } else {
+              exprCodes = inputVar.exprCode :: exprCodes
+            }
+          }
+        case _ =>
+      }
+    }
+    inputVars.toSeq
+  }
+
+  /**
+   * Helper function to calculate the size of an expression as function parameter.
+   */
+  def calculateParamLength(ctx: CodegenContext, input: Expression): Int = {
+    ctx.javaType(input.dataType) match {
+      case (ctx.JAVA_LONG | ctx.JAVA_DOUBLE) if !input.nullable => 2
```

Comment:
sorry for the question, I am just curious, why do longs and doubles require an additional param?

Reply:
It is based on the spec: https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.3

> The total length is calculated by summing the contributions of the individual parameters, where a parameter of type long or double contributes two units to the length and a parameter of any other type contributes one unit. 

If you ask why it requires two units, I don't know exactly the reason. But I guess it may related to the units long or double occupies in operand stack: https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-2.html#jvms-2.6.2

Reply:
thank you very much! :)